### PR TITLE
Add Symfony Image Component

### PR DESCRIPTION
**Pull request by avalanche123**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT
| Doc PR        | See https://github.com/romainneutron/symfony-docs/tree/symfony-image-component

This component is a port of Bulat Shakirzyanov's Imagine library (https://github.com/avalanche123/Imagine) into Symfony.

### CODE REVIEW

**Comment by stof** [comment]
I'm not a fan of integrating it in Symfony as is. The Symfony BC policy would forbid adding any new feature in the component after the initial release, unless we mark all interfaces as internal, due to the current architecture. And marking them as internal would also remove the BC promise for consuming code.

**Comment by fabpot** [comment]
@stof The thing is that the original library has enough people to maintain it. I would not like to see it unmaintained. It's been stable for years, why would we want to change the internal behavior? I mean, we can do that in the coming months/years, benefiting from our processes and the power of the Symfony community. This is just a first step.

Thanks @avalanche123 for making that possible.

**Comment by avalanche123** [comment]
@fabpot you're welcome, happy to see Imagine finding a new home, thank you!

**Comment by nicolas-grekas** [comment]
I'm supporting this proposal. Two questions to start:
- should make the component experimental on 3.3 so that we can change it while working on 3.4? (I'd do it with a note in the README + CHANGELOG, not in the code, too cumbersome, would you?)
- should we move the image fixtures to git LFS to not make the SF history grow to much?

**Comment by fabpot** [comment]
Marking the whole component as experimental for 3.3 looks like a good compromise to me.

**Comment by stof** [comment]
> should we move the image fixtures to git LFS to not make the SF git history grow too much? (maybe a bad idea)

I don't think we will change these fixtures often, meaning we will almost never store a new blob for them.
However, we should exclude them from the archive download, to avoid having every Symfony user downloading them for nothing when using composer.

**Comment by ausi** [comment]
Any interest in merging [contao/imagine-svg](https://github.com/contao/imagine-svg) too?  
I’m one of its maintainers.

**Comment by Seldaek** [comment]
While I'm not convinced moving standalone libs into Symfony is a good trend, I doubt I'll convince you not to do it, so I'd like to insist on one aspect which I think is important in terms of code design and quality: Please keep in mind it should work with very large images and that nothing should load the whole image in memory and do copies and so on. All gd/imagick transforms can be applied on files directly if done properly IIRC, and I'd like to avoid a repeat of https://github.com/avalanche123/Imagine/pull/479 if possible :) Maybe generating large fixture files while running the test suite would be an idea, or running those tests with a very low memory limit to make sure any mistake is caught early?

**Comment by andrerom** [comment]
Would also like to see the above from @Seldaek being solved before this is stable _(or even merged)_, a customer had this exact issue on a certain .sh platform, so I guess SensioCloud would have same issue @fabpot.

**Comment by fbourigault** [comment]
As Git LFS is not suitable for open-source projects, you can host fixtures on some CDN and have a phpunit listener that will download those fixtures on-demand with a sha256 checksum check. This solution allows to:
- save space in the main git repository.
- allow use of very large images in tests.
- allows you to enable caching on travis-ci.
- allows to run tests against a large set of images.
- is probably future-proof depending on which features would be added to the Image component.

I already used a such setup for a photo application with 500MB of pictures without any issue.

**Comment by javiereguiluz** [comment]
Please, let's keep things simple. Git LFS and CDN would only complicate things. The "problem" is quite limited and (probably) easy to solve:

* Tests/ is just 3,617,110 bytes
* Resources/ is 9,241,014 bytes ... and you can't move this to Git LFS or a CDN

Our biggest problems are:

* Resources\Adobe\CMYK\ files are massive! Do we need all of them?
* Resources\Adobe\Profile Information.pdf is 444 KB. Do we need it?
* Tests\Fixtures\cat.gif is 1 MB. Please, create a different GIF file.
* Tests\Fixtures\pixel-CMYK.jpg is 744 KB. Can we replace it with a smaller image?
* Tests\Fixtures\font\Arial.ttf is 773 KB. Can we replace it with a smaller and free font?

**Comment by nicolas-grekas** [comment]
@ausi that could be possible, but that'd require a license and copyright change. Would you agree on that? Anyway, I suggest to wait for this PR to be accepted and merged, then you could make a PR to add contao/imagine-svg if you'd agree to.

**Comment by ausi** [comment]
@nicolas-grekas changing the copyright and license should be no problem. 
OK, then I wait until this PR gets merged.

**Comment by sstok** [comment]
Can't we use the same approach like the old ICU component? A separate repository for the data, you already have Git so that shouldn't be a problem.

**Comment by romainneutron** [comment]
Hi everybody,

I think we're now ready, and I'm calling for reviews :)
I've addressed the issues you pointed 
 - moved the large fixtures to a dedicated repo
 - removed unnecessary ICC profiles
 - fixed the test suite (it should be alright now, no Image tests are failing)
 - I've pushed a doc branch mentioned in the PR header that contain a commit of the Imagine documentation adapted to Symfony docs. It definitely needs some work, but I'm pretty sure it would help any doc writer.
 - The whole component is marked as EXPERIMENTAL

I may have forgot something. Looking forward to read your comments.

File: src/Symfony/Component/Image/Gmagick/Image.php
Comment (stof) [inline src/Symfony/Component/Image/Gmagick/Image.php]: We may want to provide a BC layer with Imagine though, or to make a new Imagine release updating the option naming and triggering deprecation and then asking people migrating from Imagone to symfony/image to update to the latest Imagine version first

File: src/Symfony/Component/Image/Gmagick/Image.php
Comment (stof) [inline src/Symfony/Component/Image/Gmagick/Image.php]: In the method below, this is written as ``getImageHistogram``. IMO, we should update the whole component to use the "official" casing for methods

File: src/Symfony/Component/Image/Gmagick/Image.php
Comment (stof) [inline src/Symfony/Component/Image/Gmagick/Image.php]: Accessing a private static variable must always use `self`, not `static`, as child classes cannot access them.
This is not critical here (as the class is final, there cannot be a child class when calling this method), but it is still better to avoid such issue.

File: src/Symfony/Component/Image/Gmagick/Image.php
Comment (stof) [inline src/Symfony/Component/Image/Gmagick/Image.php]: same issue here

File: src/Symfony/Component/Image/Image/AbstractImage.php
Comment (stof) [inline src/Symfony/Component/Image/Image/AbstractImage.php]: BC layer of Imagine 0.x should be removed IMO

Comment (stof) [inline src/Symfony/Component/Image/Image/AbstractImage.php]: hmm, we may need to add deprecation warnings in Imagine though, as users might not know it is an old way.

File: src/Symfony/Component/Image/Image/Palette/Color/RGB.php
Comment (stof) [inline src/Symfony/Component/Image/Image/Palette/Color/RGB.php]: this short description must be removed, leaving ``Performs checks for color validity (array of array(R, G, B))`` as the actual short description

File: src/Symfony/Component/Image/Image/Palette/Color/RGB.php
Comment (stof) [inline src/Symfony/Component/Image/Image/Palette/Color/RGB.php]: this should also validate that each value is an integer in the 0 – 255 range

File: src/Symfony/Component/Image/README.md
Comment (stof) [inline src/Symfony/Component/Image/README.md]: The readme is incomplete here. It should have the basic info we have in other components

File: src/Symfony/Component/Image/Tests/Constraint/IsImageEqual.php
Comment (stof) [inline src/Symfony/Component/Image/Tests/Constraint/IsImageEqual.php]: this won't work on PHPUnit 6, as the non-namespaced class does not exist anymore

File: src/Symfony/Component/Image/Tests/Draw/AbstractDrawerTest.php
Comment (stof) [inline src/Symfony/Component/Image/Tests/Draw/AbstractDrawerTest.php]: Use ``assertFileExists``

File: src/Symfony/Component/Image/Tests/Issues/Issue59Test.php
Comment (stof) [inline src/Symfony/Component/Image/Tests/Issues/Issue59Test.php]: Tests should be renamed, as referencing Imagine issue numbers does not make sense

File: src/Symfony/Component/Image/Tests/Draw/AbstractDrawerTest.php
Comment (stof) [inline src/Symfony/Component/Image/Tests/Draw/AbstractDrawerTest.php]: you should either use a path outside the repo (in the system temp dir) or mark this folder as ignored.

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by romainneutron** [comment]
I think I've addressed all your comments. I've added a BC layer with deprecation for all options (reolution-* => resolution_*, quality, filters, resampling-filter)

File: src/Symfony/Component/Image/Image/Palette/Color/RGB.php
Comment (romainneutron) [inline src/Symfony/Component/Image/Image/Palette/Color/RGB.php]: addressed

File: src/Symfony/Component/Image/README.md
Comment (romainneutron) [inline src/Symfony/Component/Image/README.md]: addressed

File: src/Symfony/Component/Image/Gmagick/Image.php
Comment (romainneutron) [inline src/Symfony/Component/Image/Gmagick/Image.php]: addressed

**Comment by teohhanhui** [comment]
If Symfony *really* wants to introduce an Image component, it'd be great if it uses [libvips](http://www.vips.ecs.soton.ac.uk/index.php?title=VIPS). PHP binding here: https://github.com/jcupitt/php-vips

A wonderful library based on libvips in the Node.js world: https://github.com/lovell/sharp

**Comment by robfrawley** [comment]
Note that we may want to work on the brittle aspects of the unit tests before introducing this component: the Travis CI [tests break regularly](https://travis-ci.org/avalanche123/Imagine/builds) as the various extensions mature due to slightly different result values as well as deprecated and/or removed methods/constants/etc. Can we do something to mitigate these issues before adding this to the Symfony test runtime?

**Pull request closed without merging**
